File: internal/api/formatify/pyfmt.cpp
#include <iostream>
#include <string>
#include "internal.h"
#include "pyfmt.h"

static PyFmt pyfmt;

size_t PyfmtDesiredSize(const char* mode, const char* data, size_t indent) {
	std::string result = pyfmt.dumps(mode, data, indent);
	if (result.length() == 0) {
		return 0;
	}
	return result.length() + 1;
}

size_t PyfmtDumps(const char* mode, char* data, size_t size, size_t indent) {
	std::string result = pyfmt.dumps(mode, data, indent);
	if (result.length() >= size) {
		std::cout << "[CGO-error] size of result too long. ("
					<< result.length() << " >= " <<  size << ")"
					<< std::endl;
		return -1;
	}

	std::strcpy(data, result.data());
	return 0;
}

File: internal/api/formatify/pyfmt.py
import traceback
from typing import Optional
from formatify import Command
from formatify import AstText, JsonText


def dumps(mode: str, data: str, indent: int, has_escape: bool) -> Optional[str]:
    try:
        if mode == "json":
            return JsonText(data).dumps(indent, has_escape)
        elif mode == "python":
            return AstText(data).dumps(indent)
        elif mode == "command":
            return Command(data).dumps(indent)
        else:
            print(f"[PYFMT] Unsupport {mode} mode.")
            return None
    except:
        tb_info = traceback.format_exc()
        print(f"[PYFMT] dumps> {tb_info}")
        return None
